### PR TITLE
[BugFix]iceberg partition pruning support cast

### PR DESCRIPTION
**Pull request by zombee0**
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java [hunk 1]
```diff
@@ -249,14 +259,87 @@ private static Object convertBoolLiteralValue(Object literalValue) {
...
+        public Object visit(ScalarOperator scalarOperator, Type.TypeID context) {
             return null;
         }
 
         @Override
-        public Object visitConstant(ConstantOperator operator, Void context) {
+        public Object visitConstant(ConstantOperator operator, Type.TypeID context) {
+            if (context != null && needCast(operator.getType().getPrimitiveType(), context)) {
+                try {
+                    switch (context) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:304]: abstract this block to a method

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 15 / 40 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|15|40|37.50%|[272, 274, 276, 281, 291, 304, 306, 307, 309, 310, 312, 313, 315, 316, 318, 319, 322, 323, 325, 326, 333, 337, 339, 340, 341]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java [hunk 2]
```diff
@@ -249,14 +259,84 @@ private static Object convertBoolLiteralValue(Object literalValue) {
...
+                case DATE:
+                    return dstTypeID != Type.TypeID.DATE;
+                case DATETIME:
+                    return dstTypeID != Type.TypeID.TIMESTAMP;
+                default:
+                    return false;
+            }
+        }
+
+        private ConstantOperator tryCastToResultType(ConstantOperator operator, Type.TypeID resultTypeID) {
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:295]: maybe you could implement `visitCastOperator`

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java [hunk 3]
```diff
@@ -118,6 +119,14 @@ public Types.StructType getSchema() {
 
     private static class IcebergExprVisitor extends ScalarOperatorVisitor<Expression, IcebergContext> {
 
+        private static Type.TypeID getResultTypeID(String columnName, IcebergContext context) {
+            if (context == null) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:123]: context could be null?

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java [hunk 4]
```diff
@@ -249,14 +259,84 @@ private static Object convertBoolLiteralValue(Object literalValue) {
...
+                    return dstTypeID != Type.TypeID.DECIMAL;
+                case HLL:
+                case VARCHAR:
+                case CHAR:
+                    return dstTypeID != Type.TypeID.STRING;
+                case DATE:
+                    return dstTypeID != Type.TypeID.DATE;
+                case DATETIME:
+                    return dstTypeID != Type.TypeID.TIMESTAMP;
+                default:
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:290]: ARRAY,MAP,STRUCT？

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java (hunk 3 above)
Comment (zombee0) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:123]: will change it to preconditin check

File: fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java (hunk 4 above)
Comment (zombee0) [reply fe/fe-core/src/main/java/com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java:290]: These type usually won't use as partition column, ignore is ok, but i will give a better way.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 16 / 29 (55.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|16|29|55.17%|[125, 126, 264, 266, 270, 272, 274, 279, 290, 297, 302, 311, 319]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 28 / 42 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|28|42|66.67%|[135, 230, 272, 274, 276, 281, 291, 299, 303, 311, 328, 329, 330, 331]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 28 / 42 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|28|42|66.67%|[135, 230, 272, 274, 276, 281, 291, 299, 303, 311, 328, 329, 330, 331]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 34 / 42 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|34|42|80.95%|[135, 230, 281, 299, 303, 312, 331, 332]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 34 / 42 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|34|42|80.95%|[135, 230, 281, 299, 303, 312, 331, 332]|

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 34 / 42 (80.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/iceberg/ScalarOperatorToIcebergExpr.java|34|42|80.95%|[135, 230, 281, 299, 303, 312, 331, 332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by stephen-shelby — APPROVED** [review]

**Comment by zombee0** [comment]
https://github.com/Mergifyio backport branch-3.1

**Comment by zombee0** [comment]
https://github.com/Mergifyio backport branch-3.2

**Pull request merged**